hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAirdropHandler.java

```diff
@@ -0,0 +1,567 @@
+                TransactionBody.newBuilder().cryptoTransfer(body).build();
+        final var transferContext = new TransferContextImpl(context, body, true);
+        CryptoTransferExecutor.chargeCustomFee(syntheticCryptoTransferTxn, transferContext);
+    }
+
+    /**
+     *  Create new {@link AccountPendingAirdrop} and if the sender has already existing pending airdrops
+     *  link them together and update the account store and the pending airdrop store with the new values
+     */
+    private AccountPendingAirdrop getNewAccountPendingAirdropAndUpdateStores(
+            Account senderAccount,
+            PendingAirdropId pendingId,
+            PendingAirdropValue pendingValue,
+            WritableAccountStore accountStore,
+            WritableAirdropStore pendingStore) {
+
+        AccountPendingAirdrop newAccountPendingAirdrop;
+        if (senderAccount.hasHeadPendingAirdropId()) {
+            // Get the previous head pending airdrop and update the previous airdrop ID
+            var headAirdropId = senderAccount.headPendingAirdropIdOrThrow();
+            var headAccountPendingAirdrop = pendingStore.getForModify(headAirdropId);
+            validateTrue(headAccountPendingAirdrop != null, INVALID_TOKEN_ID);
+            var updatedAirdrop = headAccountPendingAirdrop
+                    .copyBuilder()
+                    .previousAirdrop(pendingId)
+                    .build();
+            pendingStore.patch(headAirdropId, updatedAirdrop);
+
+            // Create new account airdrop with next airdrop ID the previous head airdrop
+            newAccountPendingAirdrop = createAccountPendingAirdrop(pendingId, pendingValue, headAirdropId);
+        } else {
+            newAccountPendingAirdrop = AirdropHandlerHelper.createAccountPendingAirdrop(pendingId, pendingValue);
+        }
+        // Update the sender account with new head pending airdrop
+        var updatedSenderAccount =
+                senderAccount.copyBuilder().headPendingAirdropId(pendingId).build();
+        accountStore.put(updatedSenderAccount);
+        return newAccountPendingAirdrop;
+    }
+
+    /**
+     *  Calculate the fees for the token airdrop transaction. The fees are calculated by combining the
+     *  default airdrop fees, the crypto transfer fees and the token association fees.
+     */
+    @NonNull
+    @Override
+    public Fees calculateFees(@NonNull final FeeContext feeContext) {
+        final var op = feeContext.body().tokenAirdropOrThrow();
+        final var tokensConfig = feeContext.configuration().getConfigData(TokensConfig.class);
+        validateTrue(tokensConfig.airdropsEnabled(), NOT_SUPPORTED);
+
+        final var defaultAirdropFees =
+                feeContext.feeCalculatorFactory().feeCalculator(SubType.DEFAULT).calculate();
+        final var cryptoTransferFees = calculateCryptoTransferFees(feeContext, op.tokenTransfers());
+        final var tokenAssociationFees = calculateTokenAssociationFees(feeContext, op);
```

Comment:
These should be charged after the work is done. So they should be done in handle, mostly when we dispatch those transactions. If we create a pending record, why do we still charge `CryptoTransferFees` ?

Reply:
The idea of pre-paying is to have as little friction as possible. If a user accepts an airdrop, it should be immediately transferred(it won't matter if the sender doesn't have any funds). So if we have an airdrop, we know it's paid and can be accepted.

This is how it's described in the HIP and the design doc(https://github.com/hashgraph/hedera-services/blob/cae6fe99fda07b02bf003f82e3ee93436eb73a30/hedera-node/docs/design/services/token-service/airdrops/token-airdrop-transaction.md)